langchain/src/callbacks/handlers/tracer_langchain.ts

```diff
@@ -0,0 +1,207 @@
+      // eslint-disable-next-line no-process-env
+      this.headers["x-api-key"] = process.env?.LANGCHAIN_API_KEY;
+    }
+
+    this.tenantId =
+      tenantId ??
+      (typeof process !== "undefined"
+        ? // eslint-disable-next-line no-process-env
+          process.env?.LANGCHAIN_TENANT_ID
+        : undefined);
+    this.exampleId = exampleId;
+  }
+
+  async newSession(sessionName?: string): Promise<TracerSession> {
+    const tenantId = this.tenantId ?? (await this.updateTenantId());
+    const sessionCreate: TracerSessionCreateV2 = {
+      start_time: Date.now(),
+      name: sessionName,
+      tenant_id: tenantId,
+    };
+    const session = await this.persistSession(sessionCreate);
+    this.session = session;
+    return session as TracerSession;
+  }
+
+  async loadSession(sessionName: string): Promise<TracerSession> {
+    const endpoint = `${this.endpoint}/sessions?name=${sessionName}`;
+    return this._handleSessionResponse(endpoint);
+  }
+
+  async loadDefaultSession(): Promise<TracerSession> {
+    const endpoint = `${this.endpoint}/sessions?name=default`;
+    return this._handleSessionResponse(endpoint);
+  }
+
+  protected async persistSession(
+    sessionCreate: Optional<BaseTracerSessionV2, "tenant_id">
+  ): Promise<TracerSession> {
+    const endpoint = `${this.endpoint}/sessions`;
+    const tenant_id = this.tenantId ?? (await this.updateTenantId());
+    const response = await fetch(endpoint, {
+      method: "POST",
+      headers: this.headers,
+      body: JSON.stringify(sessionCreate),
+    });
+    if (!response.ok) {
+      if (sessionCreate.name !== undefined) {
+        return await this.loadSession(sessionCreate.name);
+      } else {
+        return await this.loadDefaultSession();
+      }
+    }
+    return {
+      id: (await response.json()).id,
+      tenant_id,
+      ...sessionCreate,
+    };
+  }
+
+  async updateTenantId(): Promise<string> {
+    const endpoint = `${this.endpoint}/tenants`;
+    const response = await fetch(endpoint, {
+      method: "GET",
+      headers: this.headers,
+    });
+
+    if (!response.ok) {
+      throw new Error(
+        `Failed to fetch tenant ID: ${response.status} ${response.statusText}`
+      );
+    }
+
+    const tenants = await response.json();
+    if (!tenants || tenants.length === 0) {
+      throw new Error(`No tenants found for endpoint ${endpoint}`);
+    }
+
+    const tenantId = tenants[0].id;
+    this.tenantId = tenantId;
+    return tenantId;
+  }
+
+  protected async _handleSessionResponse(
+    endpoint: string
+  ): Promise<TracerSession> {
+    const tenantId = this.tenantId ?? (await this.updateTenantId());
+    const configured_endpoint = `${endpoint}&tenant_id=${this.tenantId}`;
+    const response = await fetch(configured_endpoint, {
+      method: "GET",
+      headers: this.headers,
+    });
+    let tracerSession: TracerSession;
+    if (!response.ok) {
+      throw new Error(
```

Comment:
I removed creating a fake session if it fails, as the fake session won't be accepted by the backend anyway, so there's no point

Reply:
Should we add a `try / catch ` in `manager.ts` so the program doesn't crash when the env var is set?

Reply:
We already have that